reporters/ClickhouseReporter.js

```diff
@@ -0,0 +1,217 @@
+        description String,
+        arguments Array(String)
+    ) ENGINE = MergeTree()
+    PRIMARY KEY(crawlId, pageId, callId)`,
+    `CREATE TABLE IF NOT EXISTS ${DB}.apiCallStats (
+        crawlId String,
+        pageId String,
+        source String,
+        stats String
+    ) ENGINE = MergeTree()
+    PRIMARY KEY(crawlId, pageId, source)`,
+    `CREATE TABLE IF NOT EXISTS ${DB}.cookies (
+        crawlId String,
+        pageId String,
+        cookieId UInt32,
+        cookie String
+    ) ENGINE = MergeTree()
+    PRIMARY KEY(crawlId, pageId, cookieId)`,
+    `CREATE TABLE IF NOT EXISTS ${DB}.targets (
+        crawlId String,
+        pageId String,
+        targetId UInt32,
+        url String,
+        type String
+    ) ENGINE = MergeTree()
+    PRIMARY KEY(crawlId, pageId, targetId)`,
+];
+
+class ClickhouseReporter extends BaseReporter {
+
+    id() {
+        return 'clickhouse';
+    }
+
+    /**
+     * @param {{verbose: boolean, startTime: Date, urls: number, logPath: string, proxyHost?: string}} options 
+     */
+    init(options) {
+        this.createCrawl('', options.proxyHost || '');
+        if (options.verbose) {
+            console.log(`Creating crawl ${this.crawlId}`);
+        }
+    }
+
+    /**
+     * @param {string} name
+     * @param {string} region
+     */
+    createCrawl(name = '', region='') {
+        this.client = new ClickHouse({url: CLICKHOUSE_SERVER});
+        this.crawlId = `${new Date().toISOString()}-${os.hostname()}`;
+        this.ready = Promise.all(TABLE_DEFINITIONS.map(stmt => this.client.query(stmt).toPromise()))
+        .then(async () => {
+            await this.client.insert(`INSERT INTO ${DB}.crawls (crawlId, name, region)`, [{
+                crawlId: this.crawlId,
+                name,
+                region,
+            }]).toPromise();
+            this.queue = {
+                pages: [],
+                requests: [],
+                elements: [],
+                cmps: [],
+                apiSavedCalls: [],
+                apiCallStats: [],
+                cookies: [],
+                targets: [],
+            };
+        });
+        return this.ready;
+    }
+
+    /**
+     * Called whenever site was crawled (either successfully or not)
+     * 
+     * @param {{site: string, failures: number, successes: number, urls: number, data: import('../crawler').CollectResult, crawlTimes: Array<Array<number>>, fatalError: Error, numberOfCrawlers: number, regionCode: string}} data 
+     */
+    update(data) {
+        this.processSite(data.data);
+    }
+
+    /**
+     * @param {import('../crawler').CollectResult} data
+     */
+    processSite(data) {
+        // @ts-ignore
+        const pageId = createUrlHash(new URL(data.initialUrl));
+        this.ready = this.ready.then(async () => {
+            this.queue.pages.push([this.crawlId, pageId, data.testStarted, data.testFinished, data.initialUrl, data.finalUrl, data.timeout]);
+            if (data.data.requests) {
```

Comment:
this is not future proof (e.g. it won't pass data from new collectors to clickhouse) which is fine, but you said we can just send clickhouse JSON strings and it will be able to still work whit them, do you think it's worth doing that for unknown collectors?

Reply:
I would leave it up to the collector implementor if they want to write a Clickhouse schema and import. I wanted to avoid inserting data for some collectors (e.g. trace) as they generate a lot of data and I didn't have a use-case for working with it.